### PR TITLE
[Issue 3115] Support mining beneficiary transitions

### LINKED ISSUE #3115: Bft mining beneficiary to be configurable using transitions
Make the mining beneficiary configurable using the transitions mechanism.

**Acceptance Criteria**
- Mining beneficiary can be changed at a specific block for IBFT2
- Mining beneficiary can be changed at a specific block for QBFT

### PR DESCRIPTION
**Pull request by mbaxter**
## PR description
Add the ability to transition the `miningbeneficiary` for *bft (ibft2, qbft) consensus mechanisms. 

Example genesis file with `miningbeneficiary` transitions:
```
{
  "config": {
    "chainId": 999,
    "byzantiumBlock": 0,
    "ibft2": {
      "blockperiodseconds": 1,
      "epochlength": 30000,
      "requesttimeoutseconds": 5,
      "blockreward": "5000000000000000000",
      "miningbeneficiary": "0x0000000000000000000000000000000000000001"
    },
    "transitions": {
      "ibft2": [
          {
              "block": 10000, 
              "miningbeneficiary": "",
          },
          {
              "block": 20000, 
              "miningbeneficiary": "0x0000000000000000000000000000000000000002",
          }
      ]
    }
  },
  ...
}
```
The `miningbeneficiary` will only be updated if there is a "miningbeneficiary" key in the `transitions.ibft2` array element.  Setting the `miningbeneficiary` to an empty value will clear out any override so that block rewards go to the miner rather than a global override address.  

Other changes:
* Keep the miningbeneficiary as an `Address` rather than a `String` in `BftConfigOptions` and make the conversion from `String` to `Address` strict (i.e. all 20 address bytes are required "0x01" is not valid). 
* Fix `ThreadBesuNodeRunner` so that it respects a node's genesis config.

## Fixed Issue(s)
Fixes #3115

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/BaseBftProtocolSchedule.java [hunk 1]
```diff
@@ -82,21 +76,18 @@ public ProtocolSchedule createProtocolSchedule(
       final BftConfigOptions config, final FeeMarket feeMarket);
 
   private ProtocolSpecBuilder applyBftChanges(
-      final BftConfigOptions configOptions,
       final ProtocolSpecBuilder builder,
+      final BftConfigOptions configOptions,
       final boolean goQuorumMode,
-      final BftExtraDataCodec bftExtraDataCodec,
-      final Optional<BigInteger> blockReward) {
```
Comment (mbaxter) [inline consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/BaseBftProtocolSchedule.java:89]: Simplify method signature: just pull the `blockReward` from `configOptions` which is already provided

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/blockchain/Amount.java [hunk 2]
```diff
@@ -49,20 +49,27 @@ public Unit getUnit() {
     return unit;
   }
 
+  public Amount add(final Amount other) {
```
Comment (mbaxter) [inline acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/blockchain/Amount.java:52]: Create `Amount.add` method and extract common logic from `Amount.subract`.

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/node/ThreadBesuNodeRunner.java [hunk 3]
```diff
@@ -161,27 +162,30 @@ public void startNode(final BesuNode node) {
...
+        .gasLimitCalculator(GasLimitCalculator.constant())
+        .pkiBlockCreationConfiguration(
+            node.getPkiKeyStoreConfiguration()
+                .map(pkiConfig -> new PkiBlockCreationConfigurationProvider().load(pkiConfig)))
+        .evmConfiguration(EvmConfiguration.DEFAULT)
+        .maxPeers(maxPeers);
+
+    node.getGenesisConfig()
+        .map(GenesisConfigFile::fromConfig)
+        .ifPresent(builder::genesisConfigFile);
```
Comment (mbaxter) [inline acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/node/ThreadBesuNodeRunner.java:186]: Set the genesisConfig if it is set on the `node`.

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/BaseBftProtocolSchedule.java [hunk 4]
```diff
@@ -106,22 +97,9 @@ private ProtocolSpecBuilder applyBftChanges(
...
         .skipZeroBlockRewards(true)
-        .blockHeaderFunctions(BftBlockHeaderFunctions.forOnchainBlock(bftExtraDataCodec));
-
-    blockReward.ifPresent(bigInteger -> builder.blockReward(Wei.of(bigInteger)));
-
-    if (configOptions.getMiningBeneficiary().isPresent()) {
-      final Address miningBeneficiary;
-      try {
-        // Precalculate beneficiary to ensure string is valid now, rather than on lambda execution.
-        miningBeneficiary = Address.fromHexString(configOptions.getMiningBeneficiary().get());
```
Comment (mbaxter) [inline consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/BaseBftProtocolSchedule.java:117]: The `miningBeneficiary` is now stored as an `Address`, so we don't have to interpret it here.

File: besu/src/main/java/org/hyperledger/besu/controller/IbftBesuControllerBuilder.java [hunk 5]
```diff
@@ -141,14 +141,14 @@ protected MiningCoordinator createMiningCoordinator(
...
 
     final Address localAddress = Util.publicKeyToAddress(nodeKey.getPublicKey());
-    final BftBlockCreatorFactory blockCreatorFactory =
-        new BftBlockCreatorFactory(
+    final BftBlockCreatorFactory<?> blockCreatorFactory =
+        new BftBlockCreatorFactory<>(
             transactionPool.getPendingTransactions(),
             protocolContext,
             protocolSchedule,
+            forksSchedule,
```
Comment (mbaxter) [inline besu/src/main/java/org/hyperledger/besu/controller/IbftBesuControllerBuilder.java:149]: Pass `forkSchedule` to block creator so we can calculate the mining beneficiary based on the configured forks.

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/blockcreation/BftBlockCreator.java [hunk 6]
```diff
@@ -46,22 +49,27 @@ public BftBlockCreator(
...
-        miningBeneficiary,
         minBlockOccupancyRatio,
         parentHeader);
     this.bftExtraDataCodec = bftExtraDataCodec;
   }
 
+  private static MiningBeneficiaryCalculator miningBeneficiaryCalculator(
+      final Address localAddress, final ForksSchedule<? extends BftConfigOptions> forksSchedule) {
+    return blockNum ->
+        forksSchedule.getFork(blockNum).getValue().getMiningBeneficiary().orElse(localAddress);
```
Comment (mbaxter) [inline consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/blockcreation/BftBlockCreator.java:70]: Use the mining beneficiary from the current fork, or else default to the current block creator.

File: config/src/test/java/org/hyperledger/besu/config/BFTForkTest.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+  public void getMiningBeneficiary_fromEmptyConfig() {
+    ObjectNode config = JsonUtil.objectNodeFromMap(Collections.emptyMap());
+
+    final BftFork bftFork = new BftFork(config);
+    assertThat(bftFork.getMiningBeneficiary()).isEmpty();
+  }
+
+  @Test
+  public void getMiningBeneficiary_withValidAddress() {
+    final String beneficiary = "0x1111111111111111111111111111111111111111";
```
Comment (macfarla) [inline config/src/test/java/org/hyperledger/besu/config/BFTForkTest.java:40]: might be clearer to have a constant for this beneficiary maybe VALID_BENEFICIARY

File: config/src/test/java/org/hyperledger/besu/config/BFTForkTest.java [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+  public void getMiningBeneficiary_withInvalidValue() {
+    // Address is only 19 bytes
+    final String beneficiary = "random";
+    testGetMiningBeneficiaryWithInvalidAddress(beneficiary);
+  }
+
+  @Test
+  public void getMiningBeneficiary_withInvalidAddress() {
+    // Address is only 19 bytes
+    final String beneficiary = "0x11111111111111111111111111111111111111";
```
Comment (macfarla) [inline config/src/test/java/org/hyperledger/besu/config/BFTForkTest.java:80]: would it be clearer to do VALID_BENEFICIARY.substring(0,18)

File: config/src/test/java/org/hyperledger/besu/config/BFTForkTest.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        JsonUtil.objectNodeFromMap(
+            Map.of(BftFork.MINING_BENEFICIARY_KEY, "\t" + beneficiary + "  "));
+
+    final BftFork bftFork = new BftFork(config);
+    assertThat(bftFork.getMiningBeneficiary().map(Address::toHexString)).contains(beneficiary);
+  }
+
+  @Test
+  public void getMiningBeneficiary_withInvalidValue() {
+    // Address is only 19 bytes
```
Comment (macfarla) [inline config/src/test/java/org/hyperledger/besu/config/BFTForkTest.java:72]: this comment is not relevant here

**Review by jframe — APPROVED** [review]
Thanks implementing this, not sure when we were going to get to adding support for mining beneficiary transitions.

LGTM

**Pull request merged**